### PR TITLE
Add Easy New File Extension

### PR DESCRIPTION
**Pull request by koinzhang**
## Description

**Command**

- **New File Here**: Create a new file in the current Finder window or on the desktop.
  - New File with Name: Create a new file with name.

- **Add File Template**: Add a file template that you can use in the "New File Here" command.

**Preferences**

- Show File Template
  - [x] Document
  - [x] Code
  - [x] Script

🌟**Important Tips**:

- When you use the **New File Here** command, please activate the Finder folder windows or Desktop window where you want to create the new file.

  - You can activate it with a mouse click or switch to the Finder window with a shortcut key.
  - If no Finder window is activated, the file will be created on the desktop by default.

- When you use the **Add File Template** command, the extension will automatically detect the selected file in the frontmost Finder window.

  When using **Add File Template** command, how to get the file path automatically?

  1.  Keep your Finder window at the frontmost
  2. Select a file. (Folders will not be detected for now)
  3. Optional: You can also fill in the file directory manually.

⚠️**Caution**: This extension does not create a file when a file with the same name exists.

------

**Support File Type**

- Document
  - Plain Text
  - RTF
  - Markdown
  - Word
  - PowerPoint
  - Excel
- Code
  - Java
  - JavaScript
  - TypeScript
  - Json
  - XML
  - CSS
  - Html
- Script
  - Apple
  - Perl
  - Python
  - Ruby
  - Shell

## Screencast

https://user-images.githubusercontent.com/36128970/161391189-453d94e4-2df5-4507-91e7-3f254067d1cd.mp4

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed in the `metadata` folder

### CODE REVIEW

File: extensions/new-file-here/src/new-file-here.tsx [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+export default function main() {
+  const preference = preferences();
+  useEffect(() => {
+    async function _fetchBuildInShortcut() {
+      await runAppleScript("");
+    }
+
+    _fetchBuildInShortcut().then();
+  }, []);
```
Comment (FezVrasta) [inline extensions/new-file-here/src/new-file-here.tsx:18]: What's this needed for?

File: extensions/new-file-here/src/new-file-here.tsx [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    />
+  );
+}
+
+export function createFileName(fileType: FileType) {
+  const date = new Date();
+  const time =
+    date.getFullYear() +
+    "" +
+    date.getMonth() +
```
Comment (FezVrasta) [inline extensions/new-file-here/src/new-file-here.tsx:85]:
```suggestion
    (date.getMonth() + 1) +
```

Comment (koinzhang) [reply extensions/new-file-here/src/new-file-here.tsx:85]: Thanks for the reminder, I overlooked it here, I will submit a commit to change him

**Comment by koinzhang** [comment]
> > ⚠️Caution: It will overwrite the file with the same name
> 
> Is there a way to alert user with "file already exists" message?
> 
> If not maybe we could use `touch` command to create files which will overwrite timestamp if the file already exits and create if it doesn't.
> 
> Anyways, The extension looks amazing already! Love the file icons.

hi @Aayush9029, thank you very much for your suggestion.  I have added detection of duplicate files in the new commit.

Now, when a duplicate file exists, a new file will not be created and overwritten. Instead, a Toast will pop up to alert the user that the file already exists.

SCR-20220329-e09

**Comment by Aayush9029** [comment]
@koinzhang i had a random question, where do you get your icons from? Did you make them on your own or Is there a repo you can point me towards! 

I'd love to use them for my extensions

**Comment by koinzhang** [comment]
> @koinzhang i had a random question, where do you get your icons from? Did you make them on your own or Is there a repo you can point me towards!
> 
> I'd love to use them for my extensions

Hi@Aayush9029.

The extended icons were made by myself using the Mac icon material on this website ([macosicons](https://macosicons.com/#/)).
There are a lot of rounded rectangular icons for MAC applications on this site. I downloaded the ICNS format images from it, converted them to PNG format, and then used Figma and the [icon templates ](https://www.figma.com/community/file/1030764827259035122)provided by Raycast to create the extension icons.

**Comment by pernielsentikaer** [comment]
Hi @koinzhang 👋 

Works as expected, instead of overwriting files it just adds a count to the file name, is that correct? Since I can see it previously you ask the user?

- In the `Add File Template` I felt really lost on what it was and how to - maybe a `Form.Description` on top could help explain it a bit?

Request a new review when you are ready. Feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by koinzhang** [comment]
Hi @pernielsentikaer👋. Thank for your review and advice!

> Works as expected, instead of overwriting files it just adds a count to the file name, is that correct? Since I can see it previously you ask the user?

That is correct. Just like creating a folder on a mac system with right click, it will add a count instead of overwriting the folder.

> In the Add File Template I felt really lost on what it was and how to - maybe a Form.Description on top could help explain it a bit?

Latest change: add a `Form.Description`: The template file is equivalent to the file snippets, which you can use it to create new files directly, or just copy it.

Please continue to review.

**Comment by pernielsentikaer** [comment]
Hi @koinzhang 👋 

Thanks for taking your time to review my latest comments, I left a few more:

- There is a delay before `Template` is loaded compared to `Document` when opening `New File Here`, is there a way to first show the list when all content is ready?

- I'm not sure the `Failure Toast` when entering the `Add File Template` makes much sense since the user might not even know what the error means.

Even I have a bit of a struggle understanding when this is a smart thing and what to use it for 😄 Do you have some good use cases?

Request a new review when you are ready. Feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by koinzhang** [comment]
> There is a delay before Template is loaded compared to Document when opening New File Here, is there a way to first show the list when all content is ready?

Update: The list will now be displayed when all content is ready.

> I'm not sure the Failure Toast when entering the Add File Template makes much sense since the user might not even know what the error means.

Update:  `Failure Toast` is now only displayed when the user actively detects the selected file.


> Even I have a bit of a struggle understanding when this is a smart thing and what to use it for 😄 Do you have some good use cases?

If you are asking about the usefulness of this extension, here is my explanation:
On Windows, right click can quickly create files, but on MacOS, right click only creates folders. So creating new files quickly with the help of Raycast is a good option.
The inspiration for making this extension came from [iRightMouse](https://www.google.com.hk/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwiwitzY-aL3AhVPxIsBHU4bC1wQFnoECBkQAQ&url=https%3A%2F%2Fapps.apple.com%2Fcn%2Fapp%2F%25E8%25B6%2585%25E7%25BA%25A7%25E5%258F%25B3%25E9%2594%25AE-irightmouse%2Fid1497428978%3Fmt%3D12&usg=AOvVaw3UAXhv-yzkW9TAM0KvJzjh) and [Easy New File](https://www.google.com.hk/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwiprIOa-aL3AhW5yYsBHerxA8YQFnoECAgQAQ&url=https%3A%2F%2Fapps.apple.com%2Fus%2Fapp%2Feasy-new-file%2Fid1162194131%3Fmt%3D12&usg=AOvVaw3HK84WI5f6PUH5nas2UKBW), using them you can quickly create new files.
For example, if I want to create a new Markdown file and type some content, before I had this extension I had to open Typora and then use cmd+n to create a new file and then type the content; now I just use this extension to create a new MD file and it will open automatically and I can type the content immediately.


If you are asking what the file templates do. I added this command because the built-in file templates may not meet the needs of some users, for example, some users want to create new .ai files or other format files, they can add .ai file templates and next time they can easily create new .ai files.

**Comment by koinzhang** [comment]
Hi @pernielsentikaer 👋. Thank you for review.

> There is a delay before Template is loaded compared to Document when opening New File Here, is there a way to first show the list when all content is ready?

Update: The list will now be displayed when all content is ready.

> I'm not sure the Failure Toast when entering the Add File Template makes much sense since the user might not even know what the error means.

Update: Failure Toast is now only displayed when the user actively detects the selected file.

> Even I have a bit of a struggle understanding when this is a smart thing and what to use it for 😄 Do you have some good use cases?

If you are asking about the usefulness of this extension, here is my explanation:
On Windows, right click can quickly create files, but on MacOS, right click only creates folders. So creating new files quickly with the help of Raycast is a good option.
The inspiration for making this extension came from [iRightMouse](https://www.google.com.hk/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwiwitzY-aL3AhVPxIsBHU4bC1wQFnoECBkQAQ&url=https%3A%2F%2Fapps.apple.com%2Fcn%2Fapp%2F%25E8%25B6%2585%25E7%25BA%25A7%25E5%258F%25B3%25E9%2594%25AE-irightmouse%2Fid1497428978%3Fmt%3D12&usg=AOvVaw3UAXhv-yzkW9TAM0KvJzjh) and [Easy New File](https://www.google.com.hk/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwiprIOa-aL3AhW5yYsBHerxA8YQFnoECAgQAQ&url=https%3A%2F%2Fapps.apple.com%2Fus%2Fapp%2Feasy-new-file%2Fid1162194131%3Fmt%3D12&usg=AOvVaw3HK84WI5f6PUH5nas2UKBW), using them you can quickly create new files.
For example, if I want to create a new Markdown file and type some content, before I had this extension I had to open Typora and then use cmd+n to create a new file and then type the content; now I just use this extension to create a new MD file and it will open automatically and I can type the content immediately.

If you are asking what the file templates do. I added this command because the built-in file templates may not meet the needs of some users, for example, some users want to create new .ai files or other format files, they can add .ai file templates and next time they can easily create new .ai files.

**Pull request merged**